### PR TITLE
Submit and assert a single parameter value

### PR DESCRIPTION
**Pull request by GernotMaier**
This PR is a follow up PR of #490: submit and validate a single parameter (#490 adds code to validate tabled data).

Input data can be given in a simple yml file, e.g.:
```
---
name: ref_long
value: [-17.8920302]
units: [deg]
...
```
or as the testing file [tests/resources/reference_position_mercator.yml](tests/resources/reference_position_mercator.yml). Schema are given in the usual format (see e.g., [tests/resources/reference_position_mercator.schema.yml](tests/resources/reference_position_mercator.schema.yml).

Note the simplification of the approach in the code by converting input data to an astropy table (in this case with one row and one column). So all validation routines are identical to #490, most changes introduces here are about reading the yaml file and converting the data to tabled data.

The input data is compared to the schema, validated for units and range, and then written as a astropy table to disk (by default: ecsv). The latter is something which is probably preliminary, as we want to write model parameters into the database.

Note the addition of code to find the schema files:
- `schema` configuration parameter is a file: this is the schema file
- `schema` configuration parameter is a directory: read parameter name from data file and search for `schema/<parameter name>.schema.yml`. This will simplify workflows and points towards the intended file naming for schema files.

### CODE REVIEW

File: simtools/data_model/validate_data.py [hunk 1]
```diff
@@ -65,14 +70,68 @@ def validate_data_file(self):
...
+            self.data_table.meta["name"] = self.data["name"]
+        except KeyError as exc:
+            raise KeyError("Data dict does not contain a 'name' or 'value' key.") from exc
+
+        if self._reference_data_columns is not None:
+            self._validate_data_columns()
+
+    def _validate_data_table(self):
+        """
+        Validate tabled data.
```
Comment (orelgueta) [inline simtools/data_model/validate_data.py:110]: "tabled data" reads a bit strange. Should it be "tabulated data" or just "table data"? Google doesn't give many results for "tabled data" so I think it's not just me.

File: simtools/data_model/validate_data.py [hunk 2]
```diff
@@ -337,7 +396,7 @@ def _check_range(self, col_name, col_min, col_max, range_type="allowed_range"):
             range columns
 
         """
-        self._logger.debug(f"Checking data in column '{col_name}' for '{range_type}'")
+        self._logger.debug("Checking data in column '%s' for '%s' ", col_name, range_type)
```
Comment (orelgueta) [inline simtools/data_model/validate_data.py:399]: `f-string`

File: simtools/data_model/validate_data.py [hunk 3]
```diff
@@ -358,7 +417,9 @@ def _check_range(self, col_name, col_min, col_max, range_type="allowed_range"):
             ):
                 raise ValueError
         except KeyError:
-            self._logger.error(f"Invalid range ('{range_type}') definition for column '{col_name}'")
+            self._logger.error(
```
Comment (orelgueta) [inline simtools/data_model/validate_data.py:420]: `f-string`

File: simtools/data_model/validate_data.py [hunk 4]
```diff
@@ -466,13 +531,18 @@ def _get_reference_data_column(self, column_name, status_test=False):
...
+            _entry = [item for item in self._reference_data_columns if item["name"] == column_name]
         if status_test:
             return len(_entry) > 0
         try:
-            return _entry[0]
+            return _entry[_index]
         except IndexError:
             self._logger.error(
-                f"Data column '{column_name}' not found in reference column definition"
+                "Data column '%s' not found in reference column definition", column_name
```
Comment (orelgueta) [inline simtools/data_model/validate_data.py:546]: `f-string`

Comment (GernotMaier) [reply simtools/data_model/validate_data.py:546]: Fixed.

File: simtools/data_model/validate_data.py [hunk 5]
```diff
@@ -419,35 +485,34 @@ def _read_validation_schema(self, schema_file):
         """
 
         _schema_dict = {}
-        _data_dict = {}
+        try:
+            if schema_file.find(".schema.yml") < 0 and parameter is not None:
+                schema_file += "/" + parameter + ".schema.yml"
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/data_model/validate_data.py:490]: Done.

File: simtools/data_model/validate_data.py (hunk 3 above)
Comment (GernotMaier) [reply simtools/data_model/validate_data.py:420]: Fixed.

File: simtools/data_model/validate_data.py (hunk 2 above)
Comment (GernotMaier) [reply simtools/data_model/validate_data.py:399]: Fixed.

File: simtools/data_model/validate_data.py [hunk 6]
```diff
@@ -65,14 +70,68 @@ def validate_data_file(self):
...
+        Validate tabled data.
+
+        """
+
+        try:
+            self._reference_data_columns = self._read_validation_schema(self._schema_file_name)[
+                0
+            ].get("table_columns", None)
+        except IndexError:
+            self._logger.error("Error reading validation schema from %s", self._schema_file_name)
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/data_model/validate_data.py:119]: Fixed.

File: simtools/data_model/validate_data.py [hunk 7]
```diff
@@ -65,14 +70,68 @@ def validate_data_file(self):
...
+            _quantities = []
+            for value, unit in zip(self.data["value"], self.data["units"]):
+                try:
+                    _quantities.append(value * u.Unit(unit))
+                except ValueError:
+                    _quantities.append(value)
+            self.data_table = Table(rows=[_quantities])
+            self.data_table.meta["name"] = self.data["name"]
+        except KeyError as exc:
+            raise KeyError("Data dict does not contain a 'name' or 'value' key.") from exc
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/data_model/validate_data.py:103]: True, missed this. Added it.

File: simtools/data_model/validate_data.py [hunk 8]
```diff
@@ -65,14 +70,68 @@ def validate_data_file(self):
...
 
-        self._logger.info(f"Reading data from {self._data_file_name}")
-        self.data_table = Table.read(self._data_file_name, guess=True, delimiter=r"\s")
+        try:
+            if self._data_file_name.find("yaml") > 0 or self._data_file_name.find("yml") > 0:
+                self.data = gen.collect_data_from_yaml_or_dict(self._data_file_name, None)
+                self._logger.info("Reading data from yaml file: %s", self._data_file_name)
+            else:
+                self.data_table = Table.read(self._data_file_name, guess=True, delimiter=r"\s")
+                self._logger.info("Reading tabled data from file: %s", self._data_file_name)
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/data_model/validate_data.py:79]: Fixed.

**Comment by GernotMaier** [comment]
This is ready for the second review. Important changes:

- general.py has a new function collect_dict_from_file which allows to read a dict from a file or from a path + filename. This removes duplicated code in validate_data and metadata_collector
- required some adjustments to tests
- adjustments to the new simplified schema for parameter schema.yml files (also updated the example in tests/resources)
- download jsonschema.yml now from github webpage (before: was in test DB; which required to make sure that this DB entry is always up-to-date). Note! for now I am reading this file from a branch in the workflows repo until [this PR](https://github.com/gammasim/workflows/pull/11) is merged. 
- no changes to filenames. It continues to be `.schema.yml`. This requires more discussions and I don't agree that we should do it (e.g., there is `.simtel.zst`)

Even this PR can be reviewed, I will not merge not after approval but wait for the PR in workflows to fix the url in the integration tests.

File: simtools/data_model/validate_data.py [hunk 9]
```diff
@@ -465,14 +464,13 @@ def _interval_check(data, axis_range, range_type):
...
         Read validation schema from file.
-        Returns 'None' in case no schema file is given.
 
         Parameters
         ----------
         schema_file: Path
             Schema file describing input data.
             If this is a directory, a filename of
-            '<par>.schema.yml' is assumed.
+            '<par>_schema.yml' is assumed.
```
Comment (orelgueta) [inline simtools/data_model/validate_data.py:473]: You changed the docstring but decided against changing the code afterwards =)

File: tests/unit_tests/data_model/test_metadata_collector.py [hunk 10]
```diff
@@ -233,7 +234,11 @@ def test_collect_schema_dict(args_dict_site, tmp_test_directory):
     with open(_tmp_parameter_file, "w") as outfile:
         yaml.dump(_tmp_parameter, outfile, default_flow_style=False)
     metadata_1.args_dict["input"] = _tmp_parameter_file
-    assert metadata_1._collect_schema_dict() == _tmp_schema
+    # compared sorted dicts, because the order of the keys is not guaranteed
+    # (mostly due to above yaml.dump)
+    assert collections.OrderedDict(
```
Comment (orelgueta) [inline tests/unit_tests/data_model/test_metadata_collector.py:239]: Are you sure you need to use `OrderedDict` here instead of just `dict`?

**Review by orelgueta — COMMENTED** [review]
Very minor comments left. I could already approve technically but I would like you to check the changes I made as well.

Regarding `.schema.yml` vs. `_schema.tml`, I like the fact that we both used the `.simtel.zst` example to make our opposing points =)

**Comment by GernotMaier** [comment]
Changes have been submitted. Let me know if there is anything else.

**Review by orelgueta — APPROVED** [review]

**Pull request merged**